Significantly speedup calculation of table summaries

robintw opened the pull request:
## 🧰 Issue
Fixes #1030 

## 🚀 Overview: 
Speed up calculation of table summaries at the beginning and end of an import by not retrieving the whole object from the database, but just retrieving the relevant fields. Tests show speedups of around 2x for metadata summaries, and 6.5x for measurement summaries - although I imagine these speed-ups will be higher for the users.

## 🤔 Reason: 
Speed up imports.

## 🔨Work carried out:

- [x] Just query the created_date field rather than the whole object
- [x] Tests pass

## Confirmations

- [x] I have chosen reviewers for my PR.
- [x] I have chosen an appropriate label for the PR, adding `interactive_review` if reviewers will need to see UI
- [x] I have extended/updated the documentation in `\docs` folder
- [x] Any database content changes (Create, Edit, Delete) are recorded in the Log/Changes tables
- [x] Any database schema changes are implemented via `alembic revision` [transitions](https://pepys-imxort.readthedocs.io/en/latest/database_migration.html#how-to-use-it-for-developers)
- [x] I have completed the mandatory sections of this document.
- [x] I have deleted any unused sections.

robintw committed:
Significantly speedup calculation of table summaries

robintw commented:
@IanMayo This all works fine, and is significantly faster. However, loading the whole object from the database had the side-effect of giving us an error if there was a mismatch between the database objects definitions and the actual database schema. We have other things in place now to detect a mismatch in terms of a missing field or a misnamed field, but we won't easily catch a field type change elsewhere. If there is a field type change then it would almost certainly lead to an error at some point, but the benefit of the checking occurring here was that it would be right at the start of an import run - and so we'd catch it before we did anything else to the database.

We have tests that check that we get errors if there are differences between the table objects and the database schema, and they are all passing except the one that checks for a difference in type. I think a difference in type is likely to be fairly unlikely - particularly given that we've done most of the major schema changes now, we're far better at using database migrations too - so I think we can delete this test and continue to merge this PR. In terms of a cost-benefit analysis, I think significantly reducing delays for the users probably beats this extra check on the database structure.

IanMayo commented:
>  I think significantly reducing delays for the users probably beats this extra check on the database structure.

That seems ok for me. 

Does the omission of this test justify adding something to the PR check-list?  Or, should we just add a unit test that does the fetch?  Aaah, we're testing this specific instance of the database - so it wouldn't test the client database.  Maybe there is some merit in testing `tracstor` though, since that close models the client database.

robintw commented:
I think any change of type that wasn't reflected in the database or dealt with via a migration would almost certainly trigger an error somewhere in the manual System Test Plan, so I think it's ok to just remove this test. I'll add a commit to do that.

IanMayo commented:
I agree that the failure would arise in the System Test Plan. But, the point that we learn about the issue is while we're (I'm) preparing a release.  The value of including the test in the unit tests is that the dev will encounter the issue in the PR that introduces the problem - both allowing early resolution and giving the dev a clue as to where the error came from (the current PR).

robintw commented:
I don't quite see how we can check this in the unit tests? The databases created in the unit tests are created from the code in the repo that is being tested - so wouldn't detect a change in the code that hadn't been updated in the database. Testing against TracStor in the unit tests raises a number of issues - like how do we securely store the credentials for TracStor on both the CI and local dev machines, we would then require internet access to run all the unit tests, and we would have to keep TracStor in a state that can be used by the unit tests. The latter is the most difficult, I think, as we could easily get out of step with migrations on TracStor as migrations would have to be applied as part of the unit tests and then downgraded. I think any unit tests should use transient databases rather than something that hangs around like TracStor.

For what it's worth, I don't think we've ever run into this issue before, and I doubt we're likely to - we don't tend to go around changing the type of fields, and if we do then we'll add a migration (and other unit tests would almost certainly fail).

robintw committed:
Remove test which checked that an error was raised if a database field had changed type and the code hadn't been updated.

This used to be checked when doing the table summaries, but that got very slow - so we've removed it from there. It is now checked through the System Test Plan.

IanMayo commented:
That's completely fine Robin. thanks for explaining so clearly. I appreciate that testing against `tracstor` is fragile.

Please feel free to remove the relevant test.

IanMayo committed:
Merge branch 'develop' into speed-up-table-summaries
